# Titanic Data Analysis

## Machine Learning Model Training

- *Goal:* Train and evaluate multiple models to predict passenger survival using the the refined feature engineering file to improve passenger survival prediction

In [ ]:
# Import relevant libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

df = pd.read_csv('Titanic_for_ML_detailed.csv')
print('Data Load')

Data Load


In [47]:
# First inspection of the new ml file
df.head()

,Survived,Pclass,Age,SibSp,Parch,Fare,FamilySize,Sex_female,Sex_male,isAlone,Embarked_C,Embarked_Q,Embarked_S,TicketGroupSize,Deck_A,Deck_B,Deck_C,Deck_D,Deck_E,Deck_F,Deck_G,Deck_T,Deck_U,Title_Master,Title_Miss,Title_Mr,Title_Mrs,Title_unknown,Fine_Age_Group_Infant,Fine_Age_Group_Toddler,Fine_Age_Group_Child,Fine_Age_Group_Teenager,Fine_Age_Group_Young_Adult,Fine_Age_Group_Adult,Fine_Age_Group_Senior,FareCategory_very_cheap,FareCategory_low,FareCategory_medium,FareCategory_high,FareCategory_luxury
0,0,3,22.0,1,0,7.2500,2,0,1,0,0,0,1,1,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,1,0,0,1,0,0,0,0
1,1,1,38.0,1,0,71.2833,2,1,0,0,1,0,0,1,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1
2,1,3,26.0,0,0,7.9250,1,1,0,1,0,0,1,1,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0
3,1,1,35.0,1,0,53.1000,2,1,0,0,0,0,1,2,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1
4,0,3,35.0,0,0,8.0500,1,0,1,1,0,0,1,1,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,1,0,0,1,0,0,0


In [48]:
# Verify the set by columns and rows
print(f'Rows: {df.shape[0]} -- Columns: {df.shape[1]}')

Rows: 891 -- Columns: 40


In [49]:
df.columns

Index(['Survived', 'Pclass', 'Age', 'SibSp', 'Parch', 'Fare', 'FamilySize',
       'Sex_female', 'Sex_male', 'isAlone', 'Embarked_C', 'Embarked_Q',
       'Embarked_S', 'TicketGroupSize', 'Deck_A', 'Deck_B', 'Deck_C', 'Deck_D',
       'Deck_E', 'Deck_F', 'Deck_G', 'Deck_T', 'Deck_U', 'Title_Master',
       'Title_Miss', 'Title_Mr', 'Title_Mrs', 'Title_unknown',
       'Fine_Age_Group_Infant', 'Fine_Age_Group_Toddler',
       'Fine_Age_Group_Child', 'Fine_Age_Group_Teenager',
       'Fine_Age_Group_Young_Adult', 'Fine_Age_Group_Adult',
       'Fine_Age_Group_Senior', 'FareCategory_very_cheap', 'FareCategory_low',
       'FareCategory_medium', 'FareCategory_high', 'FareCategory_luxury'],
      dtype='object')

In [50]:
# Splitting the dataset into X the features and y the target
X = df.drop(columns=['Survived'])
y = df['Survived']

In [ ]:
# Splitting into 80% of test data and 20% validation data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Check the shape to see if both train and both test sets have the same number of rows
print('--Quick Check--')
print(f'X_test rows {X_train.shape[0]} | y_test rows: {y_train.shape[0]}')
print(f'X_train rows: {X_test.shape[0]} | y_test rows: {y_test.shape[0]}')

--Quick Check--
X_test rows 712 | y_test rows: 712
X_train rows: 179 | y_test rows: 179


In [52]:
# Creating the first model - LogisticRegression
lr_model = LogisticRegression(max_iter=1000)

# Train the model
lr_model.fit(X_train, y_train)

print('done')

done


In [53]:
# Using the trained model to make prediction
lr_predict = lr_model.predict(X_test)
lr_accur = accuracy_score(y_test, lr_predict)

# Checking the accuracy score
print(f'Accuracy Score: {lr_accur*100:.2f}%')


Accuracy Score: 81.56%


In [ ]:
# Checking for the coefficients
lr_coeff = pd.DataFrame({'Features': X.columns, 'Score': lr_model.coef_[0]})
lr_coeff = lr_coeff.sort_values(by='Score', ascending=False)
print(lr_coeff)

                      Features     Score
27       Fine_Age_Group_Infant  0.939190
17                      Deck_E  0.825241
22                Title_Master  0.812921
25                   Title_Mrs  0.797279
6                   Sex_female  0.770007
38         FareCategory_luxury  0.765556
16                      Deck_D  0.746748
37           FareCategory_high  0.353970
28      Fine_Age_Group_Toddler  0.306564
10                  Embarked_Q  0.303298
18                      Deck_F  0.291655
32        Fine_Age_Group_Adult  0.275601
9                   Embarked_C  0.037048
4                         Fare  0.001229
8                      isAlone -0.040132
1                          Age -0.050726
14                      Deck_B -0.054681
12             TicketGroupSize -0.072322
3                        Parch -0.089779
30     Fine_Age_Group_Teenager -0.120710
23                  Title_Miss -0.122573
15                      Deck_C -0.134981
33       Fine_Age_Group_Senior -0.140949
13              

In [55]:
# Creating a new model - RandomForest
rf_model = RandomForestClassifier(random_state=42)

# Train the model
rf_model.fit(X_train, y_train)

print('done')

done


In [56]:
# Using trained Random Forest model to predict
rf_predict = rf_model.predict(X_test)
rf_accur = accuracy_score(y_test, rf_predict)

# Checking the score
print(f'Score: {rf_accur*100:.2f}%')

Score: 78.21%


In [57]:
# Checking the features
rf_features = pd.DataFrame({'Features': X.columns, 'Score': rf_model.feature_importances_})
rf_features = rf_features.sort_values(by='Score', ascending=False)
print(rf_features)

                      Features     Score
1                          Age  0.145479
4                         Fare  0.142981
6                   Sex_female  0.106606
24                    Title_Mr  0.071549
7                     Sex_male  0.069635
0                       Pclass  0.052056
12             TicketGroupSize  0.043072
21                      Deck_U  0.038464
5                   FamilySize  0.034599
2                        SibSp  0.027311
23                  Title_Miss  0.027030
25                   Title_Mrs  0.025058
3                        Parch  0.017473
32        Fine_Age_Group_Adult  0.016419
11                  Embarked_S  0.015688
38         FareCategory_luxury  0.015028
9                   Embarked_C  0.013201
31  Fine_Age_Group_Young_Adult  0.012745
37           FareCategory_high  0.011159
36         FareCategory_medium  0.010928
35            FareCategory_low  0.010728
8                      isAlone  0.009951
17                      Deck_E  0.009541
34     FareCateg

In [58]:
# Creating a third new model KNN
kn_model = KNeighborsClassifier(n_neighbors=5)

# Train the model
kn_model.fit(X_train, y_train)

print('done')

done


In [59]:
# Make prediction with the trained model
kn_predict = kn_model.predict(X_test)

# Checking for the accuracy Score
kn_accur = accuracy_score(y_test, kn_predict)

print(f'Score: {kn_accur*100:.2f}%')

Score: 70.39%


**Learning:** Created overfitting. Having 40 columns introduced too much noise for the models, resulting in an accuracy score that is lower than in the previous baseline prediction. Statistically, a strong score for this dataset lies between 82% and 86%. In the following code cells, I will drop some of the redundant dummy columns to reduce the noise and improve model focus.

In [ ]:
# Dropping dummie columns to reduce noise for ml model
columns_to_drop = ['Fine_Age_Group_Infant', 'Fine_Age_Group_Toddler',
       'Fine_Age_Group_Child', 'Fine_Age_Group_Teenager',
       'Fine_Age_Group_Young_Adult', 'Fine_Age_Group_Adult',
       'Fine_Age_Group_Senior', 'FareCategory_very_cheap', 'FareCategory_low',
       'FareCategory_medium', 'FareCategory_high', 'FareCategory_luxury']

# Creating a new variable for X with the dropped columns
X_slim = X.drop(columns=columns_to_drop)

# Checking the difference between old X and new X
print(f'Old X Shape: {X.shape[1]}')
print(f'New X Shape: {X_slim.shape[1]}')


Old X Shape: 39
New X Shape: 27


In [ ]:
# Splitting the slim dataset using straticiation to maintain class distribution
X_train_slim, X_test_slim, y_train_slim, y_test_slim = train_test_split(X_slim, y, test_size=0.2, random_state=42, stratify=y)



In [ ]:
# Creating a new LR model with the new slim data
lr_model_slim = LogisticRegression(max_iter=1000)
lr_model_slim.fit(X_train_slim, y_train_slim)

# Use trained model to make prediction
lr_model_slim_predict = lr_model_slim.predict(X_test_slim)
lr_model_slim_accur = accuracy_score(y_test_slim, lr_model_slim_predict)

# Check the accuracy score
print(f'Score: {lr_model_slim_accur*100:.2f}%')

Score: 86.03 %


In [ ]:
# Checking the coefficients
lr_model_slim_coeff = pd.DataFrame({'Features': X_slim.columns, 'Score': lr_model_slim.coef_[0]})
lr_model_slim_coeff = lr_model_slim_coeff.sort_values(by='Score', ascending=False)
print(lr_model_slim_coeff)

           Features     Score
22     Title_Master  1.086689
17           Deck_E  0.906516
16           Deck_D  0.791839
25        Title_Mrs  0.739200
6        Sex_female  0.732313
18           Deck_F  0.351202
10       Embarked_Q  0.259009
9        Embarked_C  0.055753
12  TicketGroupSize  0.053919
4              Fare  0.003169
1               Age -0.036595
3             Parch -0.068539
13           Deck_A -0.117791
23       Title_Miss -0.129348
15           Deck_C -0.137028
14           Deck_B -0.174663
2             SibSp -0.236664
8           isAlone -0.257399
20           Deck_T -0.282356
5        FamilySize -0.307433
11       Embarked_S -0.316992
26    Title_unknown -0.367735
19           Deck_G -0.664178
21           Deck_U -0.675771
0            Pclass -0.729235
7          Sex_male -0.734542
24         Title_Mr -1.331035


In [69]:
# Creating new rf model and train again the slim datas
rf_model_slim = RandomForestClassifier(random_state=42)
rf_model_slim.fit(X_train_slim, y_train_slim)

# predict with the trained model
rf_model_slim_predict = rf_model_slim.predict(X_test_slim)
rf_model_slim_accur = accuracy_score(y_test_slim, rf_model_slim_predict)

# Check for the accuracy score
print(f'Score: {rf_model_slim_accur*100:.2f}%')

Score: 77.65%


In [70]:
# Checking the features of the rf model
rf_model_slim_features = pd.DataFrame({'Features': X_slim.columns, 'Importance': rf_model_slim.feature_importances_})
rf_model_slim_features = rf_model_slim_features.sort_values(by='Importance', ascending=False)
print(rf_model_slim_features)

           Features  Importance
4              Fare    0.183324
1               Age    0.181247
24         Title_Mr    0.103088
6        Sex_female    0.094376
7          Sex_male    0.062843
0            Pclass    0.059025
12  TicketGroupSize    0.051168
21           Deck_U    0.040356
5        FamilySize    0.039118
23       Title_Miss    0.029184
2             SibSp    0.026848
25        Title_Mrs    0.019826
3             Parch    0.019004
11       Embarked_S    0.014124
9        Embarked_C    0.011932
17           Deck_E    0.010272
8           isAlone    0.009724
10       Embarked_Q    0.008444
26    Title_unknown    0.007326
16           Deck_D    0.007031
22     Title_Master    0.006337
15           Deck_C    0.005656
14           Deck_B    0.003810
13           Deck_A    0.002859
18           Deck_F    0.001508
19           Deck_G    0.001156
20           Deck_T    0.000414


In [76]:
# Import standard scaler
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train_slim)
X_test_scaled = scaler.transform(X_test_slim)

# Implement KNN model with new slim data set
knn_model_slim = KNeighborsClassifier(n_neighbors=5)
knn_model_slim.fit(X_train_scaled, y_train_slim)

# Make prediction with trained model
knn_model_slim_predict = knn_model_slim.predict(X_test_scaled)
knn_model_slim_accur = accuracy_score(y_test_slim, knn_model_slim_predict)

# Checking for the accuracy score
print(f'Score: {knn_model_slim_accur*100:.2f}%')

Score: 79.33%


**Note:** By dropping several dummy columns, the LogisticRegression model improved significantly from 81.56% with 40 columns to 86.03% with 27 columns. This 4.47% increase clearly demonstrates the massiv impact of overfitting and underscores choosing the correct features is far more important than just feeding the model a large volumn data. <br>
In contrast, the Random Forest Classifier - known as a highly robust model - initially achieved an accuracy score of 78.21%, which dropped slightly to 77.65% with the slim data set. <br>
Lastly, the KNeighbors Classifier proved to be the least suited for this specific dataset, as its accuracy failed to reach 80% mark on both data versions.